From 3e465db0e8b1b2cffac34ba4d342c242b9f81f48 Mon Sep 17 00:00:00 2001
From: Rahul Bhattacharjee <quic_rbhattac@quicinc.com>
Date: Tue, 12 Jul 2022 14:47:02 +0530
Subject: [PATCH] ath11k: Fix unmapping allocated memory

In the present code we were not unmapping the allocated
memory, so in a scenerio like where we do infinite number
of FW crash and recovery, we face a crash as we run out of
memory and not able to allocate memory for futher mapping.
So, here we are cleaning up the old mappings while recovery is
happening.

Signed-off-by: Rahul Bhattacharjee <quic_rbhattac@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/core.c |  2 ++
 drivers/net/wireless/ath/ath11k/qmi.c  | 13 ++++++++++++-
 drivers/net/wireless/ath/ath11k/qmi.h  |  1 +
 3 files changed, 15 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath11k/core.c
+++ b/drivers/net/wireless/ath/ath11k/core.c
@@ -1329,6 +1329,8 @@ int ath11k_core_ssr_notifier_cb(struct n
 	ath11k_hal_dump_srng_stats(qmi->ab);
 	/* TODO Add more driver stats */
 
+	ath11k_qmi_free_target_mem_chunk(qmi->ab);
+
 	/* Crash the system once all the stats are dumped */
 	if(!qmi->ab->fw_recovery_support)
 		BUG_ON(1);
--- a/drivers/net/wireless/ath/ath11k/qmi.c
+++ b/drivers/net/wireless/ath/ath11k/qmi.c
@@ -2690,7 +2690,7 @@ out:
 	return ret;
 }
 
-static void ath11k_qmi_free_target_mem_chunk(struct ath11k_base *ab)
+void ath11k_qmi_free_target_mem_chunk(struct ath11k_base *ab)
 {
 	int i;
 
@@ -2713,6 +2713,13 @@ static void ath11k_qmi_free_target_mem_c
 		iounmap(ab->qmi.qdss_mem[0].vaddr);
 		ab->qmi.qdss_mem[0].vaddr = NULL;
 	}
+
+	if (ab->mem && ab->userpd_id) {
+		iounmap(ab->mem);
+		ab->mem = NULL;
+		ab->mem_len = 0;
+	}
+
 }
 
 static int ath11k_qmi_alloc_target_mem_chunk(struct ath11k_base *ab)
--- a/drivers/net/wireless/ath/ath11k/qmi.h
+++ b/drivers/net/wireless/ath/ath11k/qmi.h
@@ -726,5 +726,6 @@ int ath11k_enable_fwlog(struct ath11k_ba
 int ath11k_qmi_mem_read(struct ath11k_base *ab, u32 mem_addr, void *mem_value, size_t count);
 int ath11k_qmi_mem_write(struct ath11k_base *ab, u32 mem_addr, void* mem_value, size_t count);
 
+void ath11k_qmi_free_target_mem_chunk(struct ath11k_base *ab);
 
 #endif
